### PR TITLE
Snap package (edge channel) for Linux

### PR DESCRIPTION
**Pull request by pat-s**
# Implementation and installation

The following creates a usable `snap` package for every commit in the `master` branch.
This is done by doing `grunt build`  with custom settings from  `gruntfile-snap.js` to suppress unneeded building of `.deb` and other files. 

The snap is uploaded to the "edge" channel because it is built for every commit to the master branch. 

You can test the created snap beforehand by using my personal test snap `pats-boostnote`.
Simply do `snap install pats-boostnote --edge`. Then you should be able to see the application in the launcher. 
There is also a call for testing in the snap forums: https://forum.snapcraft.io/t/call-for-testing-boostnote/4085

# Stable vs edge

While the current implementation is an "edge" release, a stable release should be manually built (by simply calling `snapcraft` on the root directory in a 16.04 container to have a clean `.snap` exe) and uploaded upon every version change to the stable channel.

# Misc

I upgraded to nodejs v7 as this is the recommended version in the docs.

### CODE REVIEW

**Comment by pat-s** [comment]
WIP again, my personal travis builds are not stable. Letting you know once I have a stable build again.

**Comment by pat-s** [comment]
Completely new approach now:

A clean build is performed using `grunt`. A custom `gruntfile-snap.js` is needed so that all the unneeded `.deb` and `.rpm` files are not installed.

Note: The `grunt` plugin that is used here is a non-official one and tweaked to work for `boostnote`.
I will open a testing call for this implementation on the snap forums.

@Rokt33r Please test the app by following the instructions in the first post here.

**Comment by Rokt33r** [comment]
@pat-s 
What's the difference between gruntfile-snap.js and grundtfile.js ?

**Comment by pat-s** [comment]
`gruntfile-snap.js` only does a `grunt pre-build` instead of building all the .deb and .rpm installers as it would be done with `gruntfile.js`.
This is sufficient for the snap creation and hence reduces building time a lot.

File: gruntfile-snap.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      version: grunt.config.get('pkg.config.electron-version'),
+      'app-version': grunt.config.get('pkg.version'),
+      'app-bundle-id': 'com.maisin.boost',
+      asar: false,
+      prune: true,
+      overwrite: true,
+      out: path.join(__dirname, 'dist'),
+      ignore: /node_modules\/ace-builds\/(?!src-min)|node_modules\/ace-builds\/(?=src-min-noconflict)|node_modules\/devicon\/icons|^\/browser|^\/secret|\.babelrc|\.gitignore|^\/\.gitmodules|^\/gruntfile|^\/readme.md|^\/webpack|^\/appdmg\.json|^\/node_modules\/grunt/
+    }
+    if (platform === 'win') {
```
Comment (Rokt33r) [inline gruntfile-snap.js]: We don't need any configuration for Windows

File: gruntfile-snap.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+      })
+      packager(opts, function (err, appPath) {
+        if (err) {
+          grunt.log.writeln(err)
+          done(err)
+          return
+        }
+        done()
+      })
+    } else if (platform === 'osx') {
```
Comment (Rokt33r) [inline gruntfile-snap.js]: neither OS X

File: gruntfile-snap.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      packager(opts, function (err, appPath) {
+        if (err) {
+          grunt.log.writeln(err)
+          done(err)
+          return
+        }
+        done()
+      })
+    } else if (platform === 'linux') {
+      Object.assign(opts, {
```
Comment (Rokt33r) [inline gruntfile-snap.js]: We can just put the properties into opts without `Object.assign`

File: gruntfile-snap.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        return 'linux'
+      default:
+        return process.platform
+    }
+  }
+
+  grunt.registerTask('build', function (platform) {
+    if (platform == null) platform = getTarget()
+
+    switch (platform) {
```
Comment (Rokt33r) [inline gruntfile-snap.js]: No need to check platform. Let's just run `compile` and `pack` directly.

File: boostnote.desktop [hunk 5]
```diff
@@ -0,0 +1,8 @@
+[Desktop Entry]
+Version=9.0.0
```
Comment (Rokt33r) [inline boostnote.desktop]: It suppose to be `0.9.0`. Is it right?

And updating version manually is a little fragile. How about to create a script, which generate `boostnote.desktop` and `snap/snapcraft.yaml` with the version in `package.json`? And then, we can put the script into `postversion` of npm scripts so we can simply bump the app version with a single command, `npm version`.

**Comment by Rokt33r** [comment]
@pat-s And, could you tell me what's the differences between `gruntfile-snap.js` and `gruntfile.js`?

**Comment by Rokt33r** [comment]
If you have no time to fix, I'll try to do it in this weekend.

**Comment by pat-s** [comment]
> And, could you tell me what's the differences between gruntfile-snap.js and gruntfile.js?

See my comment above your review.

Feel free to clean up, I forget all the non-linux if statements. 

Yes, a helper scrip bumping the version might be helpful!

File: boostnote.desktop (hunk 5 above)
Comment (pat-s) [reply boostnote.desktop]: Oh right, thats a typo. As said, a helper script would be valuable yes.

**Comment by pat-s** [comment]
@Rokt33r Is there still interest in adding the `snap` or should we close here?

**Comment by Rokt33r** [comment]
@pat-s I'm sorry for the delaying...Could you give me some more specific instruction?

**Comment by Simon-the-Sorcerer** [comment]
I would like to have a distribution independent package! I think it would give the project a boost.

**Comment by benyanke** [comment]
Would love to see this as well if there was willingness to complete!

**Comment by robDAO** [comment]
What happened here guys?
@pat-s 
@benyanke
@Rokt33r

Would be very beneficial to see this snap on the stable channel!

You guys can possibly beat the competition:
Tusk
NixNote

Add-on integrations to evernote and others(zapier). Would love to see the ability to integrate with silly proprietary iCloud and pull datat from Slack. To date I am stiil not able to download notes locally from iCloud on Linux.

**Comment by pat-s** [comment]
I am not using `boostnote` anymore and won't continue here. Before, there was not really much interest. Feel free to build upon this PR :)

**Pull request closed without merging**